warehouse/attestations/services.py

```diff
@@ -0,0 +1,159 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import hashlib
+import tempfile
+
+import sentry_sdk
+
+from pydantic import TypeAdapter, ValidationError
+from pypi_attestations import (
+    Attestation,
+    AttestationType,
+    Distribution,
+    VerificationError,
+)
+from sigstore.verify import Verifier
+from zope.interface import implementer
+
+from warehouse.attestations.errors import AttestationUploadError
+from warehouse.attestations.interfaces import IAttestationsService
+from warehouse.attestations.models import Attestation as DatabaseAttestation
+from warehouse.metrics.interfaces import IMetricsService
+from warehouse.oidc.interfaces import SignedClaims
+from warehouse.oidc.models import OIDCPublisher
+from warehouse.packaging.interfaces import IFileStorage
+from warehouse.packaging.models import File
+
+
+@implementer(IAttestationsService)
+class AttestationsService:
+
+    def __init__(
+        self,
+        storage: IFileStorage,
+        metrics: IMetricsService,
+        publisher: OIDCPublisher,
+        claims: SignedClaims,
+    ):
+        self.storage: IFileStorage = storage
+        self.metrics: IMetricsService = metrics
+
+        self.publisher: OIDCPublisher = publisher
+        self.claims: SignedClaims = claims
+
+    @classmethod
+    def create_service(cls, _context, request):
+        return cls(
+            storage=request.find_service(IFileStorage),
+            metrics=request.find_service(IMetricsService),
+            claims=request.oidc_claims,
+            publisher=request.oidc_publisher,
+        )
+
+    def persist(
+        self, attestations: list[Attestation], file: File
+    ) -> list[DatabaseAttestation]:
+        attestations_db_models = []
+        for attestation in attestations:
+            with tempfile.NamedTemporaryFile() as tmp_file:
+                tmp_file.write(attestation.model_dump_json().encode("utf-8"))
+
+                attestation_digest = hashlib.file_digest(
+                    tmp_file, "blake2b"
+                ).hexdigest()
+                database_attestation = DatabaseAttestation(
+                    file=file, attestation_file_blake2_digest=attestation_digest
+                )
+
+                self.storage.store(
+                    database_attestation.attestation_path,
+                    tmp_file.name,
+                    meta=None,
+                )
+
+                attestations_db_models.append(database_attestation)
+
+        return attestations_db_models
```

Comment:
I _think_ we can do something like this instead:

```suggestion
        file.attestations.append(database_attestation)
```